core/src/main/java/org/kiwix/kiwixmobile/core/main/KiwixWebView.kt

```diff
@@ -0,0 +1,206 @@
+  }
+
+  override fun performLongClick(): Boolean {
+    val result = hitTestResult
+    if (result.type == HitTestResult.SRC_ANCHOR_TYPE) {
+      callback.webViewLongClick(result.extra)
+      return true
+    }
+    return super.performLongClick()
+  }
+
+  override fun onCreateContextMenu(menu: ContextMenu) {
+    super.onCreateContextMenu(menu)
+    val result = hitTestResult
+    if (result.type == HitTestResult.IMAGE_ANCHOR_TYPE ||
+      result.type == HitTestResult.IMAGE_TYPE ||
+      result.type == HitTestResult.SRC_IMAGE_ANCHOR_TYPE
+    ) {
+      val saveMenu =
+        menu.add(0, 1, 0, resources.getString(R.string.save_media))
+      saveMenu.setOnMenuItemClickListener {
+        val msg = SaveHandler(zimReaderContainer).obtainMessage()
+        requestFocusNodeHref(msg)
+        true
+      }
+    }
+  }
+
+  override fun onAttachedToWindow() {
+    super.onAttachedToWindow()
+    compositeDisposable.add(
+      sharedPreferenceUtil.textZooms
+        .subscribe {
+          settings.textZoom = it
+        }
+    )
+  }
+
+  override fun onDetachedFromWindow() {
+    super.onDetachedFromWindow()
+    compositeDisposable.clear()
+  }
+
+  override fun onScrollChanged(l: Int, t: Int, oldl: Int, oldt: Int) {
+    super.onScrollChanged(l, t, oldl, oldt)
+    val windowHeight = if (measuredHeight > 0) measuredHeight else 1
+    val pages = contentHeight / windowHeight
+    val page = t / windowHeight
+    callback.webViewPageChanged(page, pages)
+  }
+
+  internal class SaveHandler(private val zimReaderContainer: ZimReaderContainer) :
+    Handler() {
+    private fun getDecodedFileName(url: String?, src: String?): String {
+      var fileName = ""
+      if (url != null) {
+        fileName = url.substringAfterLast("/")
+      }
+      // If url is not a valid file name use src if it isn't null
+      if (!fileName.contains(".") && src != null) {
+        fileName = src.substringAfterLast("/")
+      }
+      return fileName.substringAfterLast("%3A")
+    }
+
+    override fun handleMessage(msg: Message) {
+      val url = msg.data["url"] as? String
+      val src = msg.data["src"] as? String
+      if (url != null || src != null) {
+        val fileName = getDecodedFileName(url, src)
+        val dotIndex = fileName.lastIndexOf('.')
+        var root =
+          Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_PICTURES)
+        if (instance.externalMediaDirs.isNotEmpty()) {
+          root = instance.externalMediaDirs[0]
+        }
+        var storageDir = File(root, fileName)
+        var newUrl = fileName
+        var i = 2
+        while (storageDir.exists()) {
+          newUrl = fileName.substring(0, dotIndex) + "_" + i + fileName.substring(dotIndex)
+          storageDir = File(root, newUrl)
+          i++
+        }
+        val source = Uri.parse(src)
+        try {
+          val input = zimReaderContainer.load("$source").data
+          storageDir.writeText(input.readBytes().toString())
```

Comment:
this is probably not the write method to be using, what this method does is save images from kiwix so the files will be much larger.

Inputstream has a kotlin method `copyTo` that I suggest you read about

Reply:
```
public fun InputStream.copyTo(out: OutputStream, bufferSize: Int = DEFAULT_BUFFER_SIZE): Long {
    var bytesCopied: Long = 0
```
@macgills It needs outputstream. How do I use it?

Reply:
explore the api of File, you should come across something

Reply:
done 

> this is probably not the write method to be using, what this method does is save images from kiwix so the files will be much larger.
 I don't understand how it affects' file size? Thanks!

Reply:
It loads the entirety of the image into memory at once as 1 big byte array. This can cause issues in a somewhat memory constrained environment like a low budget phone.

copyTo has `bufferSize: Int = DEFAULT_BUFFER_SIZE` which as you might imagine writes the file in bufferSize sized chunks